power/joules/__init__.py

```diff
@@ -0,0 +1,274 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+#
+#  hammer-vlsi plugin for Cadence Joules.
+#
+#  See LICENSE for licence details.
+
+import shutil
+from typing import List, Dict, Optional, Callable, Tuple, Set, Any, cast
+from itertools import product
+
+import os
+import errno
+import json
+
+from hammer_utils import get_or_else, optional_map, coerce_to_grid, check_on_grid, lcm_grid
+from hammer_vlsi import HammerPowerTool, HammerToolStep, MMMCCornerType, FlowLevel, TimeValue
+from hammer_logging import HammerVLSILogging
+import hammer_tech
+
+import sys
+sys.path.append(os.path.join(os.path.dirname(os.path.realpath(__file__)),"../../common"))
+from tool import CadenceTool
+
+
+class Joules(HammerPowerTool, CadenceTool):
+
+    @property
+    def post_synth_sdc(self) -> Optional[str]:
+        return None
+
+    def tool_config_prefix(self) -> str:
+        return "power.joules"
+
+    @property
+    def env_vars(self) -> Dict[str, str]:
+        new_dict = dict(super().env_vars)
+        new_dict["JOULES_BIN"] = self.get_setting("power.joules.joules_bin")
+        return new_dict
+
+    @property
+    def steps(self) -> List[HammerToolStep]:
+        return self.make_steps_from_methods([
+            self.check_level,
+            self.init_technology,
+            self.init_design,
+            self.read_stimulus,
+            self.synthesize_design,
+            self.compute_power,
+            self.report_power,
+            self.run_joules
+        ])
+
+    def check_level(self) -> bool:
+        if self.level == FlowLevel.RTL or self.level == FlowLevel.SYN:
+            return True
+        else:
+            self.logger.error("The FlowLevel is invalid. The Joules plugin only supports RTL and post-synthesis analysis. Check your power tool setting and flow step.")
+            return False
+
+    def init_technology(self) -> bool:
+        # libs, define RAMs, define corners
+        verbose_append = self.verbose_append
+
+        corners = self.get_mmmc_corners()
+        if MMMCCornerType.Extra in list(map(lambda corner: corner.type, corners)):
```

Comment:
For the future: we probably want to use `power.inputs.extra_corners_only` to sync functionality with Voltus + perhaps that option should be `true` by default. Then, this code can be cleaned up a bit.